include/swift/Runtime/SwiftDtoa.h

```diff
@@ -2,152 +2,288 @@
+/// for the binary floating point value.  The optimal form is a
+/// decimal representation that satisfies the following properties:
+///
+/// 1. Accurate.  Parsing the value back to a binary floating-point
+///    value of the same precision will exactly yield the original
+///    value.  For example, `Double(d.description) == d` for all `Double`
+///    values `d` (except for NaN values, of course).
+///
+/// 2. Short.  Of all accurate results, the returned value will
+///    contain the minimum number of significant digits.  Note that
+///    this is not quite the same as C++ `to_chars` which promises the
+///    minimal number of characters.
+///
+/// 3. Close.  Of all accurate, short results, the value printed will
+///    be the one that is closest to the exact binary floating-point
+///    value.
+///
+/// The optimal form is the ideal textual form for use in JSON and
+/// similar interchange formats because it is accurate, compact, and
+/// can be generated very quickly.  It is also ideal for logging and
+/// debugging use; the accuracy guarantees that the result can be
+/// cut-and-pasted to obtain the exact original value, and the
+/// shortness property eliminates unnecessary digits that can be
+/// confusing to readers.
+///
+/// Algorithms that produce such output have been known since at least
+/// 1990, when Steele and White published their Dragon4 algorithm.
+/// However, the earliest algorithms required high-precision
+/// arithmetic which limited their use.  Starting in 2010 with the
+/// publication of Grisu3, there has been a surge of interest and
+/// there are now a number of algorithms that can produce optimal
+/// forms very quickly.  This particular implementation is loosely
+/// based on Grisu2 but incorporates concepts from Errol and Ryu that
+/// make it significantly faster and ensure accuracy in all cases.
+///
+/// About SwiftDtoa v1
+/// ------------------
+///
+/// The first version of SwiftDtoa was committed to the Swift runtime
+/// in 2018.  It supported Swift's Float, Double, and Float80 formats.
+///
+/// About SwiftDtoa v1a
+/// -------------------
+///
+/// Version 1a of SwiftDtoa added support for Float16.
+///
+/// About SwiftDtoa v2
+/// ------------------
+///
+/// Version 2 of SwiftDtoa is a major overhaul with a number of
+/// algorithmic improvements to make it faster (especially for Float16
+/// and Double), smaller, and more portable (the code only requires
+/// C99 and makes no use of C or C++ floating-point facilities).  It
+/// also includes experimental support for IEEE 754 quad-precision
+/// binary128 format, which is not currently supported by Swift.
+//
+//===---------------------------------------------------------------------===//
 
 #ifndef SWIFT_DTOA_H
 #define SWIFT_DTOA_H
 
+// <<<<< BEGIN Local configuration overrides for Swift runtime build
+// Forcibly DISABLE binary128 support for Swift runtime
+#define SWIFT_DTOA_BINARY128_SUPPORT 0
```

Comment:
I'm tempted to define this on (or remove the condition) for the purposes of being able to back deploy a hypothetical future `Float128` type. Thoughts?

Reply:
I don't think there's any point right now.  I see three issues:

Issue:  These functions are not ABI.  Prepping for back deployment would require building out the actual ABI, including wrestling with argument-passing conventions, etc.

Issue:  We would need agreement on where to support Float128.  Should it depend on `long double` being binary128?  Will it depend on HW support?

Issue:  The float80/binary128 arithmetic is the biggest part of this formatting logic.  Turning on either one is almost 10k of code and tables, adding the second one is about 1.5k more.  So adding binary128 to platforms that already have float80 support is pretty cheap here.  But I'm wary of speculatively enabling this on platforms that don't already have float80 turned on.

Reply:
SGTM.

For the record, I can answer this:
> Issue: We would need agreement on where to support Float128. Should it depend on long double being binary128? Will it depend on HW support?

If/when we add support for Float128, we'll make it available everywhere, independent of how `long double` is defined. C provides the `_Float128` name for this purpose, so we don't need to depend on `long double` (frankly, it was a silly mistake for platforms to bind `long double` to binary128, but well, they did).